index.ts

```diff
@@ -1 +1,163 @@
+  invoiceId: string
+  items: {
+   [productCode: string]: {
+      ask: number
+      stock: number
+      unitPrice: number
+    }
+  }
+
+  constructor(invoiceId: string, sellerInfo: any, buyerInfo: any) {
+    this.invoiceId = invoiceId;
+    this.actorInfo = {
+      [Actor.Seller]: { info: sellerInfo, version: 0 },
+      [Actor.Buyer]: { info: buyerInfo, version: 0 }
+    }
+    this.items = {};
+    console.log(`Draft invoice ${invoiceId} created, seller and buyer:`, sellerInfo, buyerInfo);
+  }
+
+  isInfoOperation(op: ItemsOperation | InfoOperation): boolean {
+    // FIXME: why doesn't this work:
+    // return op instanceof ItemsOperation;
+    return ([OperationType.SetBuyerInfo, OperationType.SetSellerInfo].indexOf(op.type) !== -1);
+  }
+  
+  applyOperation (op: ItemsOperation | InfoOperation): void {
+    if (this.isInfoOperation(op)) {
+      this.applyInfoOperation(op as InfoOperation)
+    } else {
+      this.applyItemsOperation(op as ItemsOperation)
+    }
+  }
+
+  applyInfoOperation(op: InfoOperation) {
+    let actor: Actor
+    if (op.type === OperationType.SetSellerInfo) {
+      actor = Actor.Seller;
+    } else if (op.type === OperationType.SetBuyerInfo) {
+      actor = Actor.Buyer;
+    } else {
+      throw new Error(`Unknown info operation type ${op.type}`)
+    }
+    if (op.version === this.actorInfo[actor].version) {
+      throw new Error('Conflict! Same version number used twice');
+    }
+    if (op.version < this.actorInfo[actor].version) {
+      return; // old news, ignore
+    }
+    console.log(`Setting ${(actor === Actor.Buyer ? 'buyer' : 'seller')} info version ${op.version}:`, op.info);
+    this.actorInfo[actor] = { info: op.info, version: op.version };
+  }
+
+  applyItemsOperation(op: ItemsOperation) {
+    if (!this.items[op.productCode]) {
+      this.items[op.productCode] = {
+        ask: 0,
+        stock: 0,
+        unitPrice: 0
+      }
+    }
+    switch (op.type) {
+      case OperationType.AddItemQty:
+        console.log(`Adding ${op.amount} ${op.productCode} ask`);
+        this.items[op.productCode].ask += op.amount;
+        break
+      case OperationType.AddProductStock:
+        console.log(`Adding ${op.amount} ${op.productCode} stock`);
+        this.items[op.productCode].stock += op.amount;
+        break
+      case OperationType.AddProductUnitPrice:
+        console.log(`Adding ${op.amount} ${op.productCode} unit price`);
+        this.items[op.productCode].unitPrice += op.amount;
+        break
+      default:
+        throw new Error(`Unknown items operation type ${op.type}`);
+    }
+  }
+
+  allowed (op: ItemsOperation | InfoOperation, actor: Actor): boolean {
+    return (OperationsAllowed[actor].indexOf(op.type) !== -1);
+  }
+
+  processOperation(op: ItemsOperation | InfoOperation, actor: Actor): void {
+    if (this.allowed(op, actor)) {
+      console.log(`[ALLOW ${actor}]`, op);
+      return this.applyOperation(op);
+    }
+    console.log(`[DENY ${actor}]`, op);
```

Comment:
If this is enacting a local operation, this could throw "you can't do that", but ignoring it is fine I suppose.

If this is applying a remote operation, then this is catastrophic, and must trigger some coordination, otherwise the peers are guaranteed to diverge.